plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

```diff
@@ -0,0 +1,179 @@
+                                                  SemanticTokenTypes (..),
+                                                  SemanticTokens (SemanticTokens),
+                                                  UInt, absolutizeTokens)
+
+-- * 1. Mapping semantic token type to and from the LSP default token type.
+
+-- | map from haskell semantic token type to LSP default token type
+toLspTokenType :: SemanticTokenType -> Maybe SemanticTokenTypes
+toLspTokenType tk = case tk of
+  -- TVariable     -> SemanticTokenTypes_Variable
+  -- left hand side of none pattern bind
+  TFunction     -> Just SemanticTokenTypes_Function
+  TVariable     -> Just SemanticTokenTypes_Variable
+  TClass        -> Just SemanticTokenTypes_Class
+  TClassMethod  -> Just SemanticTokenTypes_Method
+  TTypeVariable -> Just SemanticTokenTypes_TypeParameter
+  -- normal data type is a tagged union type look like enum type
+  -- and a record is a product type like struct
+  -- but we don't distinguish them yet
+  TTypeCon      -> Just SemanticTokenTypes_Enum
+  TDataCon      -> Just SemanticTokenTypes_EnumMember
+  TRecField     -> Just SemanticTokenTypes_Property
+  -- pattern syn is like a limited version of macro of constructing a data type
+  TPatternSyn   -> Just SemanticTokenTypes_Macro
+  -- saturated type
+  TTypeSyn      -> Just SemanticTokenTypes_Type
+  -- not sure if this is correct choice
+  TTypeFamily   -> Just SemanticTokenTypes_Interface
+  TNothing      -> Nothing
+
+lspTokenReverseMap :: Map.Map SemanticTokenTypes SemanticTokenType
+lspTokenReverseMap = Map.fromList $ mapMaybe (\x -> fmap (,x) (toLspTokenType x)) $ enumFrom minBound
+
+fromLspTokenType :: SemanticTokenTypes -> SemanticTokenType
+fromLspTokenType tk = fromMaybe TNothing $ Map.lookup tk lspTokenReverseMap
+
+-- * 2. Mapping from GHC type and tyThing to semantic token type.
+
+toTokenType :: Name -> SemanticTokenType
+toTokenType locName = case occNameSpace $ occName locName of
+  x | isDataConNameSpace x -> TDataCon
+  x | isTvNameSpace x      -> TTypeVariable
+  x | isTcClsNameSpace x   -> TTypeCon -- Type constructors and classes in the same name space for now
+  x | isVarNameSpace x     -> TVariable
+  _                        -> TNothing
+
+-- | tyThingSemantic
+tyThingSemantic :: TyThing -> SemanticTokenType
+tyThingSemantic ty = case ty of
+  AnId vid
+    | isTyVar vid -> TTypeVariable
+    | isRecordSelector vid -> TRecField
+    | isClassOpId vid -> TClassMethod
+    | isFunVar vid -> TFunction
+    -- \| isDFunId vid -> TClassMethod
+    | otherwise -> TVariable
+  AConLike con -> case con of
+    RealDataCon _ -> TDataCon
+    PatSynCon _   -> TPatternSyn
+  ATyCon tyCon
+    | isTypeSynonymTyCon tyCon -> TTypeSyn
+    | isTypeFamilyTyCon tyCon -> TTypeFamily
+    | isClassTyCon tyCon -> TClass
+    -- fall back to TTypeCon the result
+    | otherwise -> TTypeCon
+  ACoAxiom _ -> TNothing
+  where
+    isFunVar :: Var -> Bool
+    isFunVar var = isFunType $ varType var
```

Comment:
`ghc` defines `isFunType` for you!

Reply:
We need to peels off the forall here. I don't find the ghc versions for it.

Reply:
Oh I see, you want to consider _polymorphic_ functions here. Worth a comment! You could use `GHC.dropForAlls` to get rid of the leading foralls.